include/margo-monitoring.h

```diff
@@ -0,0 +1,397 @@
+    margo_request request;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_sleep_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    double timeout_ms;
+    /* output */
+};
+
+struct margo_monitor_set_input_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t   handle;
+    margo_request request;
+    const void*   data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_set_output_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t   handle;
+    margo_request request;
+    const void*   data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_get_input_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_get_output_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_free_input_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_free_output_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_prefinalize_args {
+    margo_monitor_data_t uctx;
+};
+
+struct margo_monitor_finalize_args {
+    margo_monitor_data_t uctx;
+};
+
+struct margo_monitor_cb_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    const struct hg_cb_info* info;
+    margo_request            request;
+    /* output */
+    hg_return_t ret;
+};
+
+/**
+ * @brief Set a monitor structure for Margo to use. The structure will be
+ * internally copied and the user may free the input argument after the call.
+ * Passing NULL as monitor is valid and will disable monitor altogether.
+ *
+ * Note: if a monitor is already in place, its finalize function pointer will
+ * be called before the new monitor is installed. If provided, The initialize
+ * function of the new monitor will be called and the monitor's uargs will be
+ * replaced internally with the returned value of the initialize call.
+ *
+ * @param mid Margo instance
+ * @param monitor Monitor structure
+ * @param config JSON-formatted configuration or NULL
+ *
+ * @return HG_SUCCESS or other error code
+ */
+hg_return_t margo_set_monitor(margo_instance_id           mid,
+                              const struct margo_monitor* monitor,
+                              const char*                 config);
+
+/**
+ * @brief Invokes the on_user callback of the monitor registered with the
+ * margo instance.
+ *
+ * @param mid Margo instance
+ * @param margo_monitor_event_t Type of event
+ * @param args User-defined args for the event
+ *
+ * @return HG_SUCCESS or other error code
+ */
+hg_return_t margo_monitor_call_user(margo_instance_id mid,
```

Comment:
I really like this user-defined functionality, but the configuration here seems a little dangerous in that two separate pieces of code (the monitor, and the thing triggering this- maybe a provider?) have to agree on the type that's being passed into the void*, and there is no safety check.  I'm not sure how to solve that.  Maybe a monitor implementation defines a magic cookie value or something (`#define PHILS_MONITOR_COOKIE 23498`) and the caller here has to also provide that as an argument, then the monitor can ignore a user event that doesn't have a matching cookie value?  That's not foolproof either, but something like that might avoid obvious accidental mismatches.

Reply:
Oh I didn't think of this. Maybe a solution would be to use `const char*` instead of `void*`: whatever is provided has to be a string. If the monitoring system in place is dumping things to a trace file, then whatever string is passed is what's going to be dumped.

I initially thought we could need such a user-defined callback to add custom events with a timestamp to a trace, for very specific things, e.g. during debugging, so maybe just a `const char*` is enough, so we can provide some kind of identity to the timestamp?

Reply:
Another possibility would be to consider the Trace Event Format (https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview) which I intend to use for visualization, and make user-defined events compatible, e.g. allowing the user to provide things like a name, some categories, event type, etc.

Reply:
I think just the simple const char* approach would make sense at this level.  It's still plausible that someone might think of a reason to put a specially formatted string in there of some sort, but even so the failure mode (inability to parse) would be a graceful failure rather than a segfault.

Reply:
My latest commit changes to `const char*`.